### PR TITLE
PERF: MultiIndex.union

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v1.6.0.rst` file if fixing a bug or adding a new feature.

cc @phofl - following your improvements to MultiIndex.difference, it looks like we can replace fast_unique_multiple.

```
     [57630237]       [7d8e3cf8]
     <main>           <multiindex-union>
-      80.0±0.9ms       61.7±0.6ms     0.77  multiindex_object.SetOperations.time_operation('non_monotonic', 'string', 'union')
-        81.7±1ms         61.1±1ms     0.75  multiindex_object.SetOperations.time_operation('monotonic', 'string', 'union')
-      55.6±0.4ms       38.3±0.3ms     0.69  multiindex_object.SetOperations.time_operation('non_monotonic', 'ea_int', 'union')
-        53.8±1ms       36.6±0.8ms     0.68  multiindex_object.SetOperations.time_operation('monotonic', 'ea_int', 'union')
-      34.1±0.4ms       14.2±0.6ms     0.42  multiindex_object.SetOperations.time_operation('non_monotonic', 'int', 'union')
-        33.9±2ms       13.8±0.6ms     0.41  multiindex_object.SetOperations.time_operation('monotonic', 'int', 'union')
-        75.8±4ms       14.9±0.4ms     0.20  multiindex_object.SetOperations.time_operation('non_monotonic', 'datetime', 'union')
-        75.4±3ms       14.3±0.6ms     0.19  multiindex_object.SetOperations.time_operation('monotonic', 'datetime', 'union')
- ```

### CODE REVIEW

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -3690,10 +3690,9 @@ def _union(self, other, sort) -> MultiIndex:
             )
 
         else:
-            rvals = other._values.astype(object, copy=False)
-            right_missing = lib.fast_unique_multiple(self._values, rvals)
-            if right_missing:
-                result = self.append(other.take(right_missing))
+            right_missing = other.difference(self)
```
Comment (phofl) [inline pandas/core/indexes/multi.py:3693]: How does this behave with duplicates and especially with Indexes that contain tuples?

Apart from that, I would prefer not doing that now. I have a different pr that I would prefer merging first.

Comment (lukemanley) [reply pandas/core/indexes/multi.py:3693]: Duplicates shouldn’t reach this given the if else block this is within (other is always unique). Not sure about tuples but I can take a look. Or I can just close it if you have something in the works.

Comment (lukemanley) [reply pandas/core/indexes/multi.py]: Sure, I'll take a look at the duplicates. I'll leave this open a bit longer.

Comment (lukemanley) [reply pandas/core/indexes/multi.py]: @phofl, are you open to this one now that the duplicates case has been optimized a bit?

Comment (phofl) [reply pandas/core/indexes/multi.py]: Can you rebase now?

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx @lukemanley

**Pull request merged**